### PR TITLE
[llvm-profgen] Support COFF binary

### PR DESCRIPTION
**Pull request by HaohaiWen**
Intel Vtune/SEP has supported collecting LBR on Windows and generating
perf-script file which is same format as Linux perf script. This patch
teaches llvm-profgen to disassemble COFF binary so that we can do
Sampling based PGO on Windows.

### CODE REVIEW

**Comment by HaohaiWen** [comment]
RFC: https://discourse.llvm.org/t/llvm-profgen-support-coff-binary-for-llvm-profgen/77358

File: llvm/tools/llvm-profgen/ProfiledBinary.cpp
Comment (tru) [inline llvm/tools/llvm-profgen/ProfiledBinary.cpp:218]: ELF/COFF

File: llvm/test/tools/llvm-profgen/Inputs/coff-profile.exe
Comment (tru) [inline llvm/test/tools/llvm-profgen/Inputs/coff-profile.exe:1]: not sure we want to check in these big files into the repo, even as test data. Usually we generate binaries and PDB's with clang during the tests, but I see that there are other binaries checked in for specifically profgen. I still think it's a bit much with a exe and pdb that's almost 10MB for a single test. Maybe they can be made smaller?

**Review by tru — COMMENTED** [review]
I think this looks great! Can the current version of intel vtune already generate this file, or does it require a specific unreleased version?

I think this needs a release note, and maybe some update to llvm-profgen documentation on how it can be used with vtune for windows binaries.

File: llvm/tools/llvm-profgen/ProfiledBinary.cpp
Comment (wlei-llvm) [inline llvm/tools/llvm-profgen/ProfiledBinary.cpp:465]: Nit: assert !isCOFF() ...

File: llvm/tools/llvm-profgen/ProfiledBinary.cpp
Comment (wlei-llvm) [inline llvm/tools/llvm-profgen/ProfiledBinary.cpp:744]: Should we support `checkUseFSDiscriminator`(FS-FDO) for COFF?  the function just parses the section names to check if "__llvm_fs_discriminator__" name exists, feels it's independent to the format.

**Review by wlei-llvm — COMMENTED** [review]
Thanks for supporting the COFF binary.

File: llvm/tools/llvm-profgen/ProfiledBinary.cpp
Comment (HaohaiWen) [reply llvm/tools/llvm-profgen/ProfiledBinary.cpp]: Good caught. Thanks.

**Comment by HaohaiWen** [comment]
> I think this looks great! Can the current version of intel vtune already generate this file, or does it require a specific unreleased version?
> 
> I think this needs a release note, and maybe some update to llvm-profgen documentation on how it can be used with vtune for windows binaries.

I'm using VTune 2024.0. Here's my cmd (refer to [link](https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-0/hardware-profile-guided-optimization.html)):
```
# Open "Intel oneAPI command prompt for Intel 64 for Visual Studio 2022"
sep -start -out app.tb7 -lbr no_filter:usr -perf-script event,ip,brstack -ec BR_INST_RETIRED.NEAR_TAKEN:PRECISE=YES:SA=1000003:pdir:lbr:USR=YES -app coff-profile.exe
```
sep will generate a app.perf.data.script file. Each line of lbr records starts with "event name which triggered lbr collection" and "address"
e.g. I just deleted the prefix "BR_INST_RETIRED.NEAR_TAKEN:pdir:    7ff70a1b1482" so that llvm-profgen can directly use it.
```
BR_INST_RETIRED.NEAR_TAKEN:pdir:    7ff70a1b1482 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/-/X/A/0
```
Currently, -perf-script is not listed in sep -help but we can use it.

File: llvm/tools/llvm-profgen/ProfiledBinary.cpp
Comment (HaohaiWen) [reply llvm/tools/llvm-profgen/ProfiledBinary.cpp]: Done

File: llvm/test/tools/llvm-profgen/Inputs/coff-profile.perfscript
Comment (HaohaiWen) [reply llvm/test/tools/llvm-profgen/Inputs/coff-profile.perfscript:1]: Done

File: llvm/tools/llvm-profgen/ProfiledBinary.cpp
Comment (HaohaiWen) [reply llvm/tools/llvm-profgen/ProfiledBinary.cpp]: Done

File: llvm/test/tools/llvm-profgen/Inputs/coff-profile.exe
Comment (HaohaiWen) [reply llvm/test/tools/llvm-profgen/Inputs/coff-profile.exe:1]: I deleted the pdb file. It's safe without it.
The Symbolizer code I modified is still well tested by this lit test.

This patch make Symbolizer prefer using DWARF if it exists even if exe indicates there's PDB. We can't get mangled name from PDB so that using PDB to symbolize is meaningless for PGO (see changes https://github.com/llvm/llvm-project/pull/83972/commits/028e2476ab46af2af69576d3b8e8be493cf8a0f7).

`clang-cl.exe -O2 -gdwarf -gline-tables-only coff-profile.cpp -fuse-ld=lld -Xclang -fdebug-info-for-profiling -link -debug:dwarf`
One thing need to note is that although I had specified -debug:dwarf to lld, PDB file was still generated since clang driver inserted -debug prior to -debug:dwarf to lld linker. Therefore the coff-profile.exe contains both dwarf sections and path to PDB. This is very important to test our Symbolizer code otherwise Symbolizer will find there's no PDB binding to exe and will directly use DWARF.

If we delete the code which prefer using DWARF. The test will failed:
`llvm-profgen: error: E:\haohaiwe\compilers\llvm-project\llvm\test\tools\llvm-profgen/Inputs/coff-profile.exe: no such file or directory
`
The error message reports the exe file path and looks oddly. I've checked the code, the real reason is there's no PDB file.

Comment (HaohaiWen) [reply llvm/test/tools/llvm-profgen/Inputs/coff-profile.exe:1]: BTW, adding -gline-tables-only will reduce object size but has no obvious effect for EXE. Looks like min section size for COFF image is 4K.

**Comment by HaohaiWen** [comment]
> > I think this looks great! Can the current version of intel vtune already generate this file, or does it require a specific unreleased version?
> > I think this needs a release note, and maybe some update to llvm-profgen documentation on how it can be used with vtune for windows binaries.
> 
> I'm using VTune 2024.0. Here's my cmd (refer to [link](https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-0/hardware-profile-guided-optimization.html)):
> 
> ```
> # Open "Intel oneAPI command prompt for Intel 64 for Visual Studio 2022"
> sep -start -out app.tb7 -lbr no_filter:usr -perf-script event,ip,brstack -ec BR_INST_RETIRED.NEAR_TAKEN:PRECISE=YES:SA=1000003:pdir:lbr:USR=YES -app coff-profile.exe
> ```
> 
> sep will generate a app.perf.data.script file. Each line of lbr records starts with "event name which triggered lbr collection" and "address" e.g. I just deleted the prefix "BR_INST_RETIRED.NEAR_TAKEN:pdir: 7ff70a1b1482" so that llvm-profgen can directly use it.
> 
> ```
> BR_INST_RETIRED.NEAR_TAKEN:pdir:    7ff70a1b1482 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/P/-/A/0 0x7ff70a1b1482/0x7ff70a1b1430/-/X/A/0
> ```
> 
> Currently, -perf-script is not listed in sep -help but we can use it.


I' like to correct a mistake.
Sep is able to generate perf-script which is completely compatible with llvm-profgen.
Just specify -perf-script brstack to sep cmd.
e.g.
```
 # Open "Intel oneAPI command prompt for Intel 64 for Visual Studio 2022"
 sep -start -out app.tb7 -lbr no_filter:usr -perf-script brstack -ec BR_INST_RETIRED.NEAR_TAKEN:PRECISE=YES:SA=1000003:pdir:lbr:USR=YES -app coff-profile.exe
 ```

**Review by tru — COMMENTED** [review]
LGTM code wise.

I still think this needs some documentation and release notes. I would like to have a release note saying that it's supported in profgen and some examples on how to use in in the profgen manual page (I realize that it's not there for the ELF/perf case, but I think we can set a good precedent and document how to use this.)

**Comment by HaohaiWen** [comment]
Any more comments?
Can we land this patch?

**Review by tru — APPROVED** [review]
LGTM. But wait for another reviewer to approve before merging.

**Review by wlei-llvm — APPROVED** [review]
LGTM, thanks.

**Comment by HaohaiWen** [comment]
Thanks!

**Pull request merged**
